subsys/sched_server/cbs/cbs.c

```diff
@@ -0,0 +1,333 @@
+{
+	/*
+	 * if the budget replenishing condition is
+	 * met when a new job comes to the server,
+	 * this function is executed.
+	 */
+	cbs->abs_deadline = cycle + cbs->period;
+	cbs->budget.current = cbs->budget.max;
+
+#if defined(CONFIG_CBS_LOG) && defined(CONFIG_CBS_LOG_BUDGET_CONDITION)
+	cbs_log(CBS_BUDGET_CONDITION_MET, cbs);
+#endif
+}
+
+static void cbs_replenish_due_to_run_out(cbs_t *cbs, cbs_cycle_t cycle)
+{
+	/*
+	 * if job is still running when the budget
+	 * runs out, this function is executed.
+	 */
+	cbs->start_cycle = cycle;
+	cbs->abs_deadline += cbs->period;
+	cbs->budget.current = cbs->budget.max;
+
+#if defined(CONFIG_CBS_LOG) && defined(CONFIG_CBS_LOG_BUDGET_RAN_OUT)
+	cbs_log(CBS_BUDGET_RAN_OUT, cbs);
+#endif
+}
+
+static void cbs_budget_update_consumption(cbs_t *cbs)
+{
+	if (!cbs || !cbs->start_cycle) {
+		return;
+	}
+	cbs_cycle_t now = cbs_get_now();
+	cbs_cycle_t budget_used = now - cbs->start_cycle;
+	if (budget_used < cbs->budget.current) {
+		/*
+		 * This is the expected case when this function
+		 * is called: we have used less budget than what
+		 * we have available. A simple subtracion will
+		 * take care of updating what's left.
+		 */
+		cbs->budget.current -= budget_used;
+		return;
+	}
+	/*
+	 * If we end up here, the job has finished before
+	 * the timer expired but also spent more than
+	 * the allowed budget. We therefore need to
+	 * compensate that on replenishing.
+
+	 * It should be noted that the chances of falling
+	 * into these edge cases are higher on low-resolution
+	 * timers. The default 1ms resolution of Zephyr is
+	 * relatively low and might result in many budget
+	 * tracking inaccuracies depending on the types of
+	 * jobs executed. It is therefore recommended to
+	 * have a 100us resolution or lower instead.
+	 */
+	cbs_cycle_t excess;
+	if (budget_used > cbs->budget.max) {
+		/*
+		 * Very edge case where the job surpassed
+		 * the max allowed value for the budget
+		 * itself one or more times in a row.
+		 * It is more likely to happen when the
+		 * tick resolution is too coarse (e.g. 1ms)
+		 * and the budget is too small (e.g. 400us)
+		 */
+		for (; budget_used > cbs->budget.max; budget_used -= cbs->budget.max) {
+			cbs_replenish_due_to_run_out(cbs, cbs->start_cycle + cbs->budget.max);
+		}
+		excess = budget_used;
+	} else {
+		excess = budget_used - cbs->budget.current;
+	}
+	k_sched_lock();
+	cbs_replenish_due_to_run_out(cbs, (now - excess));
+	k_thread_deadline_set(cbs->thread, (int)(cbs->abs_deadline - cbs->start_cycle));
+	k_sched_unlock();
+	cbs->budget.current -= excess;
+}
+
+static void cbs_budget_timer_expired_callback(struct k_timer *timer)
+{
+	/*
+	 * This function is called by the timer when
+	 * it expires, which means the budget has been
+	 * entirely used and therefore needs replenishing.
+	 */
+	if (!timer) {
+		return;
+	}
+	cbs_cycle_t now = cbs_get_now();
+	cbs_t *cbs = (cbs_t *)k_timer_user_data_get(timer);
```

Comment:
FWIW if you make the k_timer a cbs field, you can just use a CONTAINER_OF() here to get the timer pointer, which is a single addition.

Reply:
great suggestion! done.